src/rustup.rs

```diff
@@ -584,73 +562,95 @@ pub fn get_rustup_version(path: &Path) -> Result<String, SyncError> {
+    download_xz: bool,
     platforms: &Platforms,
 ) -> Result<(), SyncError> {
     // Download channel file
     let channel_url = format!("{}/dist/channel-rust-{}.toml", source, channel);
     let channel_path = path.join(format!("dist/channel-rust-{}.toml", channel));
     let channel_part_path = append_to_path(&channel_path, ".part");
-    download_with_sha256_file(&channel_url, &channel_part_path, retries, true, user_agent)?;
+    download_with_sha256_file(&channel_url, &channel_part_path, retries, true, user_agent).await?;
 
     // Open toml file, find all files to download
-    let (date, files) = rustup_download_list(&channel_part_path, download_dev, &platforms)?;
-
-    // Create progress bar
-    let (pb_thread, sender) = progress_bar(Some(files.len()), prefix);
-
-    let errors_occurred = AtomicUsize::new(0);
-
-    // Download files
-    Pool::new(threads as u32).scoped(|scoped| {
-        let error_occurred = &errors_occurred;
-        for (url, hash) in &files {
-            let s = sender.clone();
-            scoped.execute(move || {
-                if let Err(e) =
-                    sync_one_rustup_target(&path, &source, &url, &hash, retries, user_agent)
-                {
-                    s.send(ProgressBarMessage::Println(format!(
-                        "Downloading {} failed: {:?}",
-                        path.display(),
-                        e
-                    )))
-                    .expect("Channel send should not fail");
-                    error_occurred.fetch_add(1, Ordering::Release);
-                }
-                s.send(ProgressBarMessage::Increment)
-                    .expect("Channel send should not fail");
+    let (date, files) = rustup_download_list(
+        &channel_part_path,
+        download_dev,
+        download_gz,
+        download_xz,
+        &platforms,
+    )?;
+    move_if_exists_with_sha256(&channel_part_path, &channel_path)?;
+
+    let pb = ProgressBar::new((files.len()) as u64)
+        .with_style(
+            ProgressStyle::default_bar()
+                .template(
+                    "{prefix} {wide_bar} {pos}/{len} [{elapsed_precise} / {duration_precise}]",
+                )
+                .progress_chars("█▉▊▋▌▍▎▏  "),
+        )
+        .with_prefix(prefix);
+    pb.enable_steady_tick(10);
+
+    let mut errors_occurred = 0usize;
+
+    let tasks = futures::stream::iter(files.iter())
+        .map(|(url, hash)| {
+            // Clone the variables that will be moved into the tokio task.
+            let path = path.to_path_buf();
+            let source = source.to_string();
+            let user_agent = user_agent.clone();
+            let url = url.clone();
+            let hash = hash.clone();
+            let pb = pb.clone();
+
+            tokio::spawn(async move {
+                pb.inc(1);
+
+                sync_one_rustup_target(&path, &source, &url, &hash, retries, &user_agent).await
             })
+        })
+        .buffer_unordered(8)
```

Comment:
Same thing here, with using the `threads` variable instead of the hardcoded value.